server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java

```diff
@@ -645,54 +652,85 @@ public Query phraseQuery(String field, TokenStream stream, int slop, boolean ena
+                    return query;
                 }
-                position += positionIncrement;
-                multiTerms.add(new Term(field, termAtt.getBytesRef()));
             }
 
-            if (enablePositionIncrements) {
-                mpqb.add(multiTerms.toArray(new Term[0]), position);
-            } else {
-                mpqb.add(multiTerms.toArray(new Term[0]));
+            if (terms.length == 1) {
+                Term[] newTerms = Arrays.stream(terms[0])
+                    .map(term -> new Term(prefixFieldType.name(), term.bytes()))
+                    .toArray(Term[]::new);
+                return new SynonymQuery(newTerms);
             }
-            return mpqb.build();
-        }
 
-        private static CachingTokenFilter cache(TokenStream in) {
-            if (in instanceof CachingTokenFilter) {
-                return (CachingTokenFilter) in;
+            SpanNearQuery.Builder spanQuery = new SpanNearQuery.Builder(name(), true);
+            spanQuery.setSlop(slop);
+            int previousPos = -1;
+            for (int i = 0; i < terms.length; i++) {
+                Term[] posTerms = terms[i];
+                int posInc = positions[i] - previousPos;
+                previousPos = positions[i];
+                if (posInc > 1) {
+                    spanQuery.addGap(posInc - 1);
+                }
+                if (i == lastPos) {
+                    if (posTerms.length == 1) {
+                        FieldMaskingSpanQuery fieldMask =
+                            new FieldMaskingSpanQuery(new SpanTermQuery(new Term(prefixFieldType.name(), posTerms[0].bytes())), name());
+                        spanQuery.addClause(fieldMask);
+                    } else {
+                        SpanQuery[] queries = Arrays.stream(posTerms)
+                            .map(term -> new FieldMaskingSpanQuery(
+                                new SpanTermQuery(new Term(prefixFieldType.name(), term.bytes())), name())
+                            )
+                            .toArray(SpanQuery[]::new);
+                        spanQuery.addClause(new SpanOrQuery(queries));
+                    }
+                } else {
+                    if (posTerms.length == 1) {
+                        spanQuery.addClause(new SpanTermQuery(posTerms[0]));
+                    } else {
+                        SpanTermQuery[] queries = Arrays.stream(posTerms)
+                            .map(SpanTermQuery::new)
+                            .toArray(SpanTermQuery[]::new);
+                        spanQuery.addClause(new SpanOrQuery(queries));
+                    }
+                }
             }
-            return new CachingTokenFilter(in);
+            return spanQuery.build();
         }
 
-        private static boolean hasGaps(CachingTokenFilter stream) throws IOException {
+        private static boolean hasGaps(TokenStream stream) throws IOException {
```

Comment:
Does this need to change?  I like it being a CachingTokenFilter just because that ensures that we have a re-usably token stream here.

Reply:
The issue I had with this is that we are building a caching token filter but we never use it outside  of the `hasGaps` function. Moreover the token filter that is passed to this function is always a caching token filter (the `QueryBuilder` wraps the token stream into a CachingTokenFilter) so I removed it, I can restore if you think it is needed.

Reply:
Maybe change it to an assertion?

Reply:
++